### PR TITLE
Decouple SaveGameFileChooser from HeadlessGameServer

### PR DESCRIPTION
**Pull request by ssoloff**
## Overview

Decouples the static methods in the `SaveGameFileChooser` class from `HeadlessGameServer`.

The `getAutoSaveFileName()` method was only ever called in a headless context, so I renamed it appropriately to make this clear.  Also note that `AUTOSAVE_TYPE#getFileName()` is only ever called in a headless context.

The remaining methods could be called in either a headed or headless context, so I added a new Boolean parameter for the caller to specify this.  My philosophy in these decoupling PRs is to keep pushing the responsibility for determining if the context is headless up the stack until they converge in hopefully one or two places where it will be easier to inject this state from `HeadlessGameServer` or `GameRunner`.

## Functional Changes

None.

## Refactoring Changes

* Renamed the `SaveGameFileChooser` methods and constants to place the adjective before the noun.
* Merged the implementations of the `getXXXAutoSaveFileName()` methods into a single method after first adding characterization tests.

## Manual Testing Performed

* **Headless game server:** Verified headless/even/odd auto-saves were created at the appropriate points in a game.
* **Headed game server:** Verified the even/odd auto-saves were created at the appropriate points in a game.

### CODE REVIEW

File: game-core/src/main/java/games/strategy/engine/framework/startup/launcher/ServerLauncher.java [hunk 1]
```diff
@@ -316,7 +316,7 @@ private void stopGame() {
   private void saveAndEndGame(final INode node) {
     // a hack, if headless save to the autosave to avoid polluting our savegames folder with a million saves
```
Comment (RoiEXLab) [inline game-core/src/main/java/games/strategy/engine/framework/startup/launcher/ServerLauncher.java]: Why am I not surprised by this comment ? 🤦‍♂️

File: game-core/src/main/java/games/strategy/engine/framework/ui/SaveGameFileChooser.java [hunk 2]
```diff
@@ -8,25 +8,41 @@
...
-  private static final String AUTOSAVE_EVEN_ROUND_FILE_NAME = GameDataFileUtils.addExtension("autosave_round_even");
+
+  @VisibleForTesting
+  static final String HEADLESS_AUTOSAVE_FILE_NAME = GameDataFileUtils.addExtension("autosave");
+  @VisibleForTesting
+  static final String ODD_ROUND_AUTOSAVE_FILE_NAME = GameDataFileUtils.addExtension("autosave_round_odd");
+  @VisibleForTesting
+  static final String EVEN_ROUND_AUTOSAVE_FILE_NAME = GameDataFileUtils.addExtension("autosave_round_even");
+
   private static SaveGameFileChooser instance;
```
Comment (RoiEXLab) [inline game-core/src/main/java/games/strategy/engine/framework/ui/SaveGameFileChooser.java]: Note, that this is unrelated to the PR, just some general information I'd like to share:

This seems like a code smell to me, there's absolutely no good reason (I could think of) to hold on to a swing object, via a static reference, that's obviously intended for one-time use only.
I looked through the usages of the getInstance methods, and they are really just using this method to display either a save or an open dialog.
Theres a lot of code going through GameRunner to set the mainFrame as parent component which is clearly not used for this dialog, only for the classic AWT dialog for mac users.
I have access to a mac at work, so I might be able to check if the thing with the AWT dialog vs Swing dialog still holds true nowadays.

TL;DR: Mac and other OSes behaviour should be unified (parent component should be mainFrame or something) and a SaveGameFileChooser instance should be created dynamically, not as a sort of singleton.

Comment (ssoloff) [reply game-core/src/main/java/games/strategy/engine/framework/ui/SaveGameFileChooser.java]: Yup, I thought the same thing. :smile:
